torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,354 @@
+r"""
+This class defines the abstraction for a Graph feature store. The goal of a
+feature store is to abstract away all node and edge feature memory management
+so that varying implementations can allow for independent scale-out.
+
+This particular feature store abstraction makes a few key assumptions:
+    * The features we care about storing are graph node and edge features. To
+        this end, the attributes that the feature store supports include a
+        group_name (e.g. a heterogeneous node name, a heterogeneous edge type,
+        etc.), an attr_name (which defines the name of the feature tensor,
+        e.g. `feat`, `discrete_feat`, etc.), and an index.
+    * A feature can be uniquely identified from any associated attributes
+        specified in :obj:`TensorAttr`.
+
+It is the job of a feature store implementor class to handle these assumptions
+properly. For example, a simple in-memory feature store implementation may
+concatenate all metadata values with a feature index and use this as a unique
+index in a KV store. More complicated implementations may choose to partition
+features in interesting manners based on the provided metadata.
+
+Major TODOs for future implementation:
+* Async `put` and `get` functionality
+"""
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Optional, Union
+
+import numpy as np
+import torch
+
+from torch_geometric.typing import FeatureTensorType
+from torch_geometric.utils.mixin import CastMixin
+
+_field_status = Enum("FieldStatus", "UNSET")
+
+IndexType = Union[FeatureTensorType, slice]
```

Comment:
```suggestion
IndexType = Union[Tensor, np.array, slice, int]
```
Looks a bit weird to use `FeatureTensor` as `IndexType` as well. Maybe make this here explicit as well: